v2.2/architecture/life-of-a-distributed-transaction.md

```diff
@@ -0,0 +1,204 @@
+
+Importantly, this feature is entirely built for transactional optimization (known as [transaction pipelining](transaction-layer.html#transaction-pipelining)). There are no issues if an operation passes the evaluator but doesn't end up committing.
+
+### Writing the transaction record
+
+The very first range that a transaction interacts with also receives a write for a key known as its transaction record. This includes a UUID to uniquely identify the transaction, as well as the canonical record of its state––e.g. `PENDING`, `ABORTED`, `COMMITTED`. This record is used by all operation to determine the state of the transaction. For example, when a write is blocked by another write, it periodically checks its own transaction record to ensure that it hasn't been aborted.
+
+As soon as the write for the transaction record completes (i.e., achieves consensus through Raft), it begins communicating with the gateway node's `TxnCoordSender`, which regularly heartbeats the transaction to keep it alive (i.e., in a `PENDING` state). If the heartbeating from the TxnCoordSender stops, the TxnRecord moves itself into `ABORTED` status.
+
+This write is handled like all other writes, which we'll discuss in more detail in a subsequent section.
+
+### Reads from RocksDB
+
+All operations (including writes) begin by reading from the local instance of RocksDB to check for write intents for the operation's key. We talk much more about [write intents in the transaction layer of the CockroachDB architecture](transaction-layer.html#write-intents), which is worth reading, but a simplified explanation is that these are provisional, uncommitted writes that express that some other concurrent transaction plans to write a value to the key.
+
+What we detail below is a simplified version of the CockroachDB transaction model. For more detail, check out [the transaction architecture documentation](transaction-layer.html).
+
+#### Resolving Write Intents
+
+If an operation encounters a write intent for a key, it attempts to "resolve" the write intent by checking the write intent's transaction's state. If the write's intent's transaction is...
+
+- `COMMITTED`, this operation converts the write intent to a regular key-value pair, and then proceeds as if it had read that value instead of a write intent.
+- `ABORTED`, this operation discards the write intent and reads the next-most-recent value from RocksDB.
+- `PENDING`, the new transaction attempts to "push" the write intent's transaction by moving that transaction's timestamp forward (i.e. ahead of this transaction's timestmap). The push succeeds only if this transaction's priority is explicitly higher; i.e. it has a `HIGH` priority or the transaction whose write intent it encountered has a `LOW` priority.
```

Comment:
It's no longer sufficient for the transaction's priority to be explicitly higher. Instead, it must either have the maximum priority or the other txn must have the minimum priority for the push to succeed. SQL-related txns will never have these extreme priorities, so they'll never result in pushes succeeding. Instead, transactions will queue up behind each other.

This is part of a shift we made back in 1.1 to move from an "optimistic" concurrency control to a "pessimistic" concurrency control. We observed that in a distributed environment where txn aborts are very expensive, OCC was too costly.